### PR TITLE
Check BTC node config without IO overhead

### PR DESCRIPTION
**Pull request by dmos62**
Edit: not proof of concept anymore; this initial post is outdated. TLDR is that this PR has Bisq check if nodes it connects to are suitable for use by Bisq (e.g. that the node isn't pruning). That's currently most useful with a local BTC node or custom nodes (because otherwise Bisq uses our own federated nodes that are known to be well configured), but it checks all BTC nodes it connects to, with negligible overhead. This PR is currently stagnating in large part because it depends on a (minor) BitcoinJ PR.

This is a proof of concept for checking BTC node's configuration. It's an alternative approach to what I had proposed earlier (that was merged and reverted due to reliability issues). In contrast to the previous attempt, this approach doesn't incur additional IO and doesn't use blocking. Whereas the previous attempt initiated a short lived BitcoinJ connection to a node, this approach just listens into the connections we are already establishing. Also, while previously the checking applied only to a local BTC node, this approach generalizes checking to all BTC nodes we connect to.

Currently, it has two problems:

a) connecting to a pruning node, on my setup at least, causes the connection to crash, and the listener never gets triggered; this is a rare BitcoinJ edge case bug and I opened an issue for it https://github.com/bisq-network/bisq/issues/4080 ;

and, b) this WIP implementation just does `peer.close()`, which causes the PeerGroup to just retry the connection, in case we're in local mode, or possibly retry later in case we're in non-local mode (didn't explore non-local behaviour yet); I'm pondering if this is indeed undesirable or if we would want to use this to get free auto-retry.

Another open question is what to do in case a remote node is found to be misconfigured. Just an error log?

### CODE REVIEW

**Comment by ripcurlx** [comment]
@dmos62 Your build is broken right now: https://travis-ci.org/github/bisq-network/bisq/builds/665235608?utm_source=github_status&utm_medium=notification

> a) connecting to a pruning node, on my setup at least, causes the connection to crash, and the listener never gets triggered; this is a rare BitcoinJ edge case bug and I opened an issue for it #4080 ;

Yes, I get following exception in the logs
```
März-23 15:46:57.444 [NioClientManager] WARN  o.b.core.PeerSocketHandler: [127.0.0.1]:18444 -  org.bitcoinj.core.ProtocolException: Received AlertMessage before version handshake is complete.
	at org.bitcoinj.core.Peer.processMessage(Peer.java:478)
	at org.bitcoinj.core.PeerSocketHandler.receiveBytes(PeerSocketHandler.java:184)
	at org.bitcoinj.net.ConnectionHandler.handleKey(ConnectionHandler.java:223)
	at org.bitcoinj.net.NioClientManager.handleKey(NioClientManager.java:86)
	at org.bitcoinj.net.NioClientManager.run(NioClientManager.java:122)
	at com.google.common.util.concurrent.AbstractExecutionThreadService$1$2.run(AbstractExecutionThreadService.java:66)
	at com.google.common.util.concurrent.Callables$4.run(Callables.java:122)
	at org.bitcoinj.utils.ContextPropagatingThreadFactory$1.run(ContextPropagatingThreadFactory.java:49)
	at java.base/java.lang.Thread.run(Thread.java:844)
```
> 
> and, b) this WIP implementation just does `peer.close()`, which causes the PeerGroup to just retry the connection, in case we're in local mode, or possibly retry later in case we're in non-local mode (didn't explore non-local behaviour yet); I'm pondering if this is indeed undesirable or if we would want to use this to get free auto-retry.

You mean a retry to the same peer?

> Another open question is what to do in case a remote node is found to be misconfigured. Just an error log?

It depends. If it would be a misconfigured provided node/public node I would just ignore it and not connect to it. If it is a node the user has configured by using a config/start parameter I think we should display a warning to the user. Also if it is a misconfigured local node we want to notify the user that something is not correct.

In general if we do this check after the connection I think it should be fine if we block the setup process at that point (e.g. when a local bitcoin node is misconfigured) and do it as with your previous implementation to allow the user to shutdown and fix this issue.

**Comment by dmos62** [comment]
>> and, b) this WIP implementation just does peer.close(), which causes the PeerGroup to just retry the connection, in case we're in local mode, or possibly retry later in case we're in non-local mode (didn't explore non-local behaviour yet); I'm pondering if this is indeed undesirable or if we would want to use this to get free auto-retry.
> 
>You mean a retry to the same peer?

Yes. I didn't double-check, but I think BitcoinJ uses a FILO queue for keeping addresses of nodes. It pops and connects to them until wanted number of connections is established. In case we're in local-mode, then that queue has size 1 and contains localhost. `peer.close()` probably causes the node to go to the back of the queue.

> block the setup process [when a local bitcoin node is misconfigured] and do it as with your previous implementation to allow the user to shutdown and fix this issue

I'd be in favor of letting an unclosable error popup be displayed asynchroniously. That would involve less fiddling with the setup process (less likely to break something).

**Comment by ripcurlx** [comment]
> I'd be in favor of letting an unclosable error popup be displayed asynchroniously. That would involve less fiddling with the setup process (less likely to break something).

I think this should be fine as well.

**Comment by dmos62** [comment]
This PR now depends on this PR on our BitcoinJ fork: https://github.com/bisq-network/bitcoinj/pull/35. For development I used a BitcoinJ build from a local Maven repository. When that PR for BitcoinJ is merged, relevant changes to `build.gradle` should be made here.

Screenshot:

user provided node misconfigured

As seen in the screenshot, we're not providing a button to revert to not using a user-provided node (don't see how that can be done easily). The reason why before we could offer that option before is that we were blocking on the check and we were only handling the local bitcoin node.

It can also be seen that Bisq is actually loaded, and after you click "shutdown" the informational local BTC node popup will appear briefly before Bisq closes.

In the background, the misconfigured peers are not explicitly closed (BitcoinJ closes them automatically if they're pruning). The errors are logged and if custom BTC nodes are used, or a local BTC node, the above error popup locks the user out.

In other words, advantage over previous tries at getting this implemented, in terms of features, is that now we get errors when remote nodes are misconfigured, and the user is locked out if the misconfigured node is local or custom. Before we were only checking the local Bitcoin node.

**Comment by dmos62** [comment]
Bump

**Comment by chimp1984** [comment]
@dmos62 Sorry that the PR got delayed so long... I asked @oscarguindzberg to review your BitcoinJ PR. If that is done I think we should get that PR in as well asap.

File: core/src/main/java/bisq/core/btc/setup/WalletsSetup.java [hunk 1]
```diff
@@ -224,11 +228,12 @@ protected void onSetupCompleted() {
                 if (preferences.getBitcoinNodes() != null && !preferences.getBitcoinNodes().isEmpty())
                     peerGroup.setAddPeersFromAddressMessage(false);
 
+                peerGroup.addVersionMessageReceivedEventListener(WalletsSetup.this::handleConfiguration);
```
Comment (chimp1984) [inline core/src/main/java/bisq/core/btc/setup/WalletsSetup.java]: We get called from the BitcoinJ thread. We should map it to our userThread to avoid threading issues.
```
peerGroup.addVersionMessageReceivedEventListener((peer, versionMessage) ->
                        UserThread.execute(handleConfiguration(peer, versionMessage)));
```

File: core/src/main/resources/i18n/displayStrings.properties [hunk 2]
```diff
@@ -2527,6 +2527,7 @@ It will make debugging easier if you include the bisq.log file by pressing "Open
 
 popup.error.tryRestart=Please try to restart your application and check your network connection to see if you can resolve the issue.
 popup.error.takeOfferRequestFailed=An error occurred when someone tried to take one of your offers:\n{0}
+popup.error.userBtcNodeMisconfigured.explanation=The user-provided Bitcoin node {0} was found to be misconfigured. A user-provided BTC node is either a locally running node that was automatically detected and used, or a node explicitly specified via the --btcNodes command-line option. Bisq requires BTC nodes to have bloom filters enabled and pruning disabled. Automatic use of a local Bitcoin node can be disabled by using the --ignoreLocalBtcNode option.
```
Comment (chimp1984) [inline core/src/main/resources/i18n/displayStrings.properties]: I think we should not show an Error popup as it comes with the bug-report text and is confusing for that use case, even it is somehow an error as Bisq cannot be used in that case. I would suggest to use warning popup and to reflect that in the display string we should use `popup.warning.userBtcNodeMisConfigured.explanation`. Also changed to camelCase as IDEA static code inspector complains about a typo otherwise. I think misconfigured should be mis-configured in the text.

File: desktop/src/main/java/bisq/desktop/main/MainViewModel.java [hunk 3]
```diff
@@ -401,6 +401,15 @@ private void setupHandlers() {
         daoPresentation.getBsqSyncProgress().addListener((observable, oldValue, newValue) -> updateBtcSyncProgress());
 
         bisqSetup.setFilterWarningHandler(warning -> new Popup().warning(warning).show());
+
+        walletsSetup.setDisplayUserBtcNodeMisconfigurationHandler(
+                (String peer) ->
+                    new Popup()
+                        .hideCloseButton()
+                        .error(Res.get("popup.error.userBtcNodeMisconfigured.explanation", peer))
```
Comment (chimp1984) [inline desktop/src/main/java/bisq/desktop/main/MainViewModel.java]: As per prev. comment:
-> ` .warning(Res.get("popup.warning.userBtcNodeMisConfigured.explanation", peer))`

File: core/src/main/java/bisq/core/btc/setup/WalletsSetup.java [hunk 4]
```diff
@@ -397,45 +402,27 @@ private void configPeerNodes(@Nullable Socks5Proxy proxy) {
...
-      return Futures.transform(versionMessageFuture, handler);
-    }
-
-    private static Void handleConfiguration(Peer peer, VersionMessage versionMessage) {
+    /* Provided a Peer and its VersionMessage, will check and handle Peer's advertised
+     * configuration. The VersionMessage is provided separately, because it is not
+     * always available and this is meant to be called from a listener that is triggered
+     * when the VersionMessage becomes available.
+     */
+    private Void handleConfiguration(Peer peer, VersionMessage versionMessage) {
```
Comment (chimp1984) [inline core/src/main/java/bisq/core/btc/setup/WalletsSetup.java]: The uppercase Void is probably a leftover from the previous code. Should be `void`.

File: core/src/main/java/bisq/core/btc/setup/WalletsSetup.java [hunk 5]
```diff
@@ -397,45 +402,27 @@ private void configPeerNodes(@Nullable Socks5Proxy proxy) {
...
-        peer.close();
+        log.error("Peer _badly_ configured (pruning and/or bloom filters disabled): {}", peer);
+        var usingLocalNode = localBitcoinNode.shouldBeUsed();
+        var usingCustomNodes =
+            BtcNodes.BitcoinNodesOption.CUSTOM.ordinal() == preferences.getBitcoinNodesOptionOrdinal();
+        if (usingLocalNode || usingCustomNodes) {
+            displayUserBtcNodeMisconfigurationHandler.accept(peer.toString());
+        }
       }
       return null;
```
Comment (chimp1984) [inline core/src/main/java/bisq/core/btc/setup/WalletsSetup.java]: If `void` is used above the return need to be removed.

**Review by chimp1984 — CHANGES_REQUESTED** [review]

File: core/src/main/resources/i18n/displayStrings.properties (hunk 2 above)
Comment (dmos62) [reply core/src/main/resources/i18n/displayStrings.properties]: Warning doesn't seem entirely appropriate, giving the event a milder air than it might deserve (event being not something to keep in mind, but something to fix), but it might be more appropriate than Error, which seems to be dedicated to events that should be bug-reported. I'll switch to Warning.

I think the spelling of "misconfigure" is correct. "Mis" + "construct" is a rarer combination, so it's not found on many dictionaries, but you can find more common uses of the prefix mis in entries like "misplace" [0] or "misconduct" [1]. If you take a look at the descriptions of the prefix mis- in popular dictionaries [2], it's normal to use it without a hyphen. I think a hyphen is used uniquely when connecting entire words (something like the hyphenation in post-modernism [3] is a discouraged spelling: redirects to postmodernism).

> IDEA static code inspector complains about a typo

Can words be added to its dictionary, like you would in a spellchecking text editor? It would be awkward to alter writing to match a spellchecker.

[0] https://www.thefreedictionary.com/misplace
[1] https://www.thefreedictionary.com/misconduct
[2] https://www.thefreedictionary.com/mis-
[3] https://www.thefreedictionary.com/post-modernism

Comment (chimp1984) [reply core/src/main/resources/i18n/displayStrings.properties]: You can set the header message as well, so the title is not warning but something more appropriate. The error type popup carries the automatic support for bug report...

Comment (chimp1984) [reply core/src/main/resources/i18n/displayStrings.properties]: Re misconfigure: Ah ok. I am not native speaker and IDEA suggestions is my teacher ;-).
> Can words be added to its dictionary
Yes.

**Review by chimp1984 — DISMISSED** [review]
utACK

Thanks for the update!

**Comment by chimp1984** [comment]
@ripcurlx This is tagged as 1.4.0 can u remove that? Can u remove also the dopped tag. It is still an important PR. I guess we are now ready from the bitcoinj side to complete it. 
@dmos62  can you check with @oscarguindzberg if there is anything missing?

**Comment by chimp1984** [comment]
@dmos62 Do you have time to continue work on that?

**Comment by dmos62** [comment]
@chimp1984 Yep. I'll get in touch with @oscarguindzberg.

**Comment by dmos62** [comment]
I've ported this PR and [a BitcoinJ PR](https://github.com/bisq-network/bitcoinj/pull/36) that this depends on to BitcoinJ 15.8. No noteworthy changes required.

This PR currently needs a locally installed BitcoinJ library instance, for development purposes (you have to `git checkout` the above mentioned BitcoinJ PR and run `gradle install` on it). Hence Travis failing. Once the BitcoinJ PR is merged, we can update this PR to use our Github hosted fork.

File: core/src/main/java/bisq/core/btc/setup/WalletsSetup.java [hunk 6]
```diff
@@ -409,6 +420,64 @@ private void configPeerNodes(@Nullable Socks5Proxy proxy) {
...
+    ///////////////////////////////////////////////////////////////////////////////////////////
+
+    /* Provided a Peer and its VersionMessage, will check and handle Peer's advertised
+     * configuration. The VersionMessage is provided separately, because it is not
+     * always available and this is meant to be called from a listener that is triggered
+     * when the VersionMessage becomes available.
+     */
+    private void handleConfiguration(Peer peer, VersionMessage versionMessage) {
+        var wellConfigured = isWellConfigured(versionMessage);
+        if (wellConfigured) {
```
Comment (kaiwolfram) [inline core/src/main/java/bisq/core/btc/setup/WalletsSetup.java:434]: A temp variable is not necessary imo, the method call is already very concise and readable
```suggestion
        if (isWellConfigured(versionMessage)) {
```

File: core/src/main/java/bisq/core/btc/setup/WalletsSetup.java [hunk 7]
```diff
@@ -409,6 +420,64 @@ private void configPeerNodes(@Nullable Socks5Proxy proxy) {
...
+            log.info("Peer well configured: {}", peer);
+        } else {
```
Comment (kaiwolfram) [inline core/src/main/java/bisq/core/btc/setup/WalletsSetup.java:436]: Redundant `else` clause bc the `if` can return. Makes it more readable
```suggestion
        if (wellConfigured) {
            log.info("Peer well configured: {}", peer);
            return;
        }
```

File: core/src/main/java/bisq/core/btc/setup/WalletsSetup.java (hunk 6 above)
Comment (dmos62) [reply core/src/main/java/bisq/core/btc/setup/WalletsSetup.java:434]: A variable defining the condition is a style I use (and propagate) and, even if the called method is descriptive as well, I still use it. One reason is consistency of style. Another, the if statement doesn't care that the node being well configured is predicated on the version message (that's an implemenation detail), so there's no reason to put that in the if statement. I would be against your suggestion, because it optimizes for terseness at the cost of readability.

File: core/src/main/java/bisq/core/btc/setup/WalletsSetup.java (hunk 7 above)
Comment (dmos62) [reply core/src/main/java/bisq/core/btc/setup/WalletsSetup.java:436]: If you don't remove the `wellConfigured` intermediary variable as you suggested above (which I am against), this originally reads "if the node is well configured log that, else perform following steps", which is favourable over your suggetion: "if the node is well configured, given provided version message, exit the method, foregoing steps that would only be executed in case the node were not well configured." Your edits have a negative effect on readability and a very minimal effect on terseness, which, in my opinion, should not be a consideration anyway (but that's besides the point).
